infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/DockerInternalRuntime.java

```diff
@@ -194,68 +200,56 @@ protected void internalStop(Map<String, String> stopOptions) throws Infrastructu
         return Collections.unmodifiableMap(properties);
     }
 
-    private void startMachine(String name,
-                              DockerContainerConfig containerConfig,
-                              Map<String, String> startOptions,
-                              boolean isDev) throws InfrastructureException {
-        DockerMachine dockerMachine;
-        // TODO property name
-        final RuntimeIdentity identity = getContext().getIdentity();
-        if ("true".equals(startOptions.get("restore"))) {
-            MachineSourceImpl machineSource;
-            try {
-                SnapshotImpl snapshot = snapshotDao.getSnapshot(identity.getWorkspaceId(),
-                                                                identity.getEnvName(),
-                                                                name);
-                machineSource = snapshot.getMachineSource();
-                // Snapshot image location has SHA-256 digest which needs to be removed,
-                // otherwise it will be pulled without tag and cause problems
-                String imageName = machineSource.getLocation();
-                if (imageName.contains("@sha256:")) {
-                    machineSource.setLocation(imageName.substring(0, imageName.indexOf('@')));
-                }
+    private void restoreMachine(String name, DockerContainerConfig config) throws InfrastructureException {
+        try {
+            SnapshotImpl snapshot = snapshotDao.getSnapshot(identity.getWorkspaceId(), identity.getEnvName(), name);
+            config = configForSnapshot(snapshot, config);
+        } catch (NotFoundException x) {
+            noSnapshotHandler.handle(name);
+        } catch (SnapshotException x) {
+            throw new InternalInfrastructureException(x);
+        }
+        startMachine(name, config);
+    }
 
-                DockerContainerConfig imageContainerConfig = normalizeSource(containerConfig, machineSource);
-                dockerMachine = serviceStarter.startService(dockerEnvironment.getNetwork(),
-                                                            name,
-                                                            imageContainerConfig,
-                                                            identity,
-                                                            isDev);
-            } catch (NotFoundException | SnapshotException | SourceNotFoundException e) {
```

Comment:
Here is one tricky case. When SourceNotFoundException is thrown we should fall back to start from origin source. Please, add this workaround.